code/game/machinery/_machinery.dm

```diff
@@ -955,11 +955,36 @@
 	return TRUE
 
 /obj/machinery/proc/display_parts(mob/user)
-	. = list()
-	. += span_notice("It contains the following parts:")
-	for(var/obj/item/C in component_parts)
-		. += span_notice("[icon2html(C, user)] \A [C].")
-	. = jointext(., "")
+	var/list/part_count = list()
+	for(var/obj/item/component_part in component_parts)
+		if(!part_count[component_part.name])
+			if(isstack(component_part))
+				var/obj/item/stack/stack_part = component_part
+				part_count[component_part.name] = stack_part.amount
+			else
+				part_count[component_part.name] = 1
+		else
+			part_count[component_part.name]++
+
+	var/count = 0
+	var/suffix = ""
+	var/list/printed_components = list()
+
+	var/text = span_notice("It contains the following parts:")
+	for(var/obj/item/component_part in component_parts)
+		if(!printed_components[component_part.name])
```

Comment:
Early-return

Reply:
Printed components is used just an temporary list to check which components were already printed for example if I have 2 matter bins then I set printed components["matter bin"]=true so the 2nd matter bin in the next iteration  won't get printed again.
If I return here then It would not print remaining components in the foor loop and would break the code

Reply:
Early-continue then

Reply:
Ummm that's would break the counting  logic completely  I'll give u an example why. 
Say I have 2 matter bins if I applied your logic and printed_components is empty 
if(!printed_components["matter_bin"])
  continue 
Then the 1st Matter bin is skipped and not printed beacause the list is empty and initially does not have this entry 
Same thing for the 2nd matter bin it's skipped and not printed so that would be a logical error

Reply:
You would invert it

```dm
if (printed_components[name])
    part_count[name]++
    continue
```

Reply:
That's correct however part_count[name]++ isn't necessary because it was already initialzed in the previous for loop and should only be printed now so it would be 

if (printed_components[name])
    continue  //already printed dint print again 

//print stuff 

I'll do that